Add support for processCreds for SIW v2

conorhanrahan-okta opened the pull request:
## Description:

This PR adds support for processCreds in an IDX sign in widget. Not every view has access to the current user's login. Some views contain only a password input, and the server response contains a user that looks like this:
```
  "user": {
    "type": "object",
    "value": {
      "id": "00uwb8GLwf1HED5Xs0g3"
    }
  },
```

If we want those views to call `processCreds`, we will need to have the backend send something like this:
```
  "user": {
    "type": "object",
    "value": {
      "id": "I9bvFiq01cRFgbn",
      "passwordChanged": "2019-05-03T19:00:00.000Z",
      "profile": {
        "login": "foo@example.com",
        "firstName": "Foo",
        "lastName": "Bar",
        "locale": "en-us",
        "timeZone": "UTC"
      }
    }
  },
```
I'm guessing the user's profile is not being sent over for security reasons.

These views will call processCreds
- IdentifierView (with a password)
- EnrollAuthenticatorPasswordView

These views will not call processCreds because the FE doesn't have access to the user's username:
- ChallengeAuthenticatorPasswordView
- ReEnrollAuthenticatorPasswordView
- ReEnrollAuthenticatorWarningPasswordView
- ResetAuthenticatorPasswordView

## PR Checklist

- [ ] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [x] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video:


### Reviewers:


### Issue:

- [OKTA-314099](https://oktainc.atlassian.net/browse/OKTA-XXXXXX)

conorhanrahan-okta commented:
Got all Selenium tests to pass on the downstream bacon: https://bacon-go.aue1e.saasure.net/commits?artifact=monolith&branch=d16t-okta-signin-widget-8fa47ee-607d04ce

sherizada-okta commented on src/v2/view-builder/views/password/EnrollAuthenticatorPasswordView.js:
```diff
@@ -73,6 +73,18 @@ const Body = BaseForm.extend({
+    }
+    processCredsPromise.then(() => {
+      BaseForm.prototype.saveForm.apply(this, arguments);
+    });
```
It looks like the Settings method `processCreds` takes care of ensuring `this.settings.processCreds` always returns a promise.

sherizada-okta replied:
And same thing about the identifer check here (let processCreds handle it either on our end of theirs)

sherizada-okta commented on src/v2/view-builder/views/IdentifierView.js:
```diff
@@ -20,28 +20,42 @@ const Body = BaseForm.extend({
+      username: this.model.get('identifier'),
+      password
+    };
+    return this.settings.processCreds(creds);
```
I think the checks for identifier/password should either
- go inside `this.settings.processCreds`
or
- not happen at all and just let the user handle it

haishengwu-okta replied:
what about avoid to invoke `this.settings.processCreds` if it's identify first flow (only has user name field)

conorhanrahan-okta replied:
That's how I did it originally in this commit: https://github.com/okta/okta-signin-widget/pull/1851/commits/8fa47ee77a4940a66d1c56ab295364a94bfec53d

After I saw I changed it to call processCreds with partial credentials and do the check inside `this.settings.processCreds`. Which approach do we want to take here? I don't feel strongly either way

haishengwu-okta replied:
Don't have strong opinion either way but leaning toward to not make changes to `this.settings.processCreds` method. hence the ask.

conorhanrahan-okta replied:
Reverted this change

sherizada-okta commented:
Can you clean up the PR title?

conorhanrahan-okta commented:
Will this need a Selenium test or is the TestCafe tests enough? Not sure if it's possible to do `widgetrc` stuff during a Selenium test

haishengwu-okta commented on src/models/Settings.js:
```diff
@@ -397,9 +397,9 @@ export default Model.extend({
   processCreds: function(creds) {
     const processCreds = this.get('processCreds');
-
+    const credentialsAreComplete = creds.username && creds.password;
```
why need add this check?

conorhanrahan-okta replied:
See the discussion here: https://github.com/okta/okta-signin-widget/pull/1851#discussion_r616153242

conorhanrahan-okta replied:
I've reverted this change

haishengwu-okta reviewed (commented):
looks like any password related View shall invoke this call? 
- in v1, we have PrimaryAuth, PasswordExpired, PasswordReset

conorhanrahan-okta commented:
> looks like any password related View shall invoke this call?
> 
> * in v1, we have PrimaryAuth, PasswordExpired, PasswordReset

The call to `processCreds` will only happen in the views listed in the description. In most of the password views we don't have access to the user's identifier, so for those views there will be no underlying call to the user's `processCreds`.

sherizada-okta commented on src/v2/view-builder/views/IdentifierView.js:
```diff
@@ -20,28 +20,38 @@ const Body = BaseForm.extend({
+    this.getDeviceFingerprintPromise()
+      .then(() => this.getProcessCredsPromise())
+      .then(() => {
+        BaseForm.prototype.saveForm.apply(this, arguments);
```
`arguments` here should be `undefined` because you're inside an arrow function

sherizada-okta commented on src/v2/view-builder/views/password/EnrollAuthenticatorPasswordView.js:
```diff
@@ -73,6 +73,15 @@ const Body = BaseForm.extend({
+      username: this.options.appState.get('user')?.profile?.login,
+      password: this.model.get('credentials.passcode')
+    }).then(() => {
+      BaseForm.prototype.saveForm.apply(this, arguments);
```
arrow functions have no `arguments`

conorhanrahan-okta replied:
I just looked into this, it looks like the arguments variable works fine here. Arrow function's don't set the `arguments` variable at all, which means it get the `arguments` value from the outer scope. I set a breakpoint to double check. The existing device fingerprinting logic takes the same approach.

sherizada-okta commented on test/testcafe/spec/EnrollAuthenticatorPasswordView_spec.js:
```diff
@@ -84,3 +88,18 @@ test('should have the correct reqiurements', async t => {
+  await rerenderWidget({
+    processCreds: ({ username, password }) => {
+      // eslint-disable-next-line no-console
+      console.log(`processCreds called with ${username} and ${password}`);
```
what about when processCreds is a promise

conorhanrahan-okta replied:
I modified one of these tests to use the `processCreds(creds, callback)` function signature instead, I am guessing that's what you were referring to. All the promise stuff happens in `this.settings.processCreds`, not the`this.settings.get('processCreds')` that the user supplies.

sherizada-okta replied:
Right, like a test case supporting this example from the [readme](https://github.com/okta/okta-signin-widget#username-and-password):
```
// Passed a creds object {username, password} and a callback for further processing
processCreds: function (creds, callback) {
  // This example demonstrates optional legacy form-based login
  $.ajax({
    method: "POST",
    url: "/logintype",
    data: {
      username : creds.username
    },
    success: function (logintype) {
      if (logintype == "LEGACY") {
        $('#legacyUser').val(creds.username);
        $('#legacyPassword').val(creds.password);
        $('#legacyLogonForm').submit();
      } else {
        callback();
      }
    }
  });
}
```

sherizada-okta commented:
Can you also add a row for this new OIE behavior here: https://oktawiki.atlassian.net/wiki/spaces/eng/pages/1894023233/OIE+Vs.+Legacy+behaviour+drift

If you search for "SIW customization" you'll get some examples. It sounds like the drift is OIE hook won't support the same flows.

conorhanrahan-okta commented:
I spent some time looking at the v1 behavior and realized a few places where I was differing from existing processCreds behavior:
- In the Identifier view, I now call processCreds with an `undefined` username when there is no password in the view. This is similar to the v1 behavior in `PrimaryAuthForm.js` when `features.passwordlessAuth` is enabled
- I realized that processCreds is not being called when enrolling a password in v1. So I deleted the commit where I added process creds to v2's `EnrollAuthenticatorPasswordView`

To summarize the new behavior:
- `processCreds` will only be enabled for the Identifier which includes a password field.
- `processCreds` will not be called when in the expired password or reset password views. This is different than the v1 behavior. If we want to support this, we will need the BE to pass the profile in with the user (see [this Slack comment](https://okta.slack.com/archives/CAZH2MWEL/p1618513936470300))

Sorry, I regret not double checking this earlier! 😖

sherizada-okta commented:
So in these instances, it's best to ask the PM what the best course of action is. Which I've done [here](https://okta.slack.com/archives/G01NZUZJCBH/p1619543746072800).

conorhanrahan-okta commented:
Closing because this will be too different from the old behavior. See https://okta.slack.com/archives/G01NZUZJCBH/p1619543746072800
